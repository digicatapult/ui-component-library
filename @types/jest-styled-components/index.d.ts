declare module 'jest-styled-components/serializer'
