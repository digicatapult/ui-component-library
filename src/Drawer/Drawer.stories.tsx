import React from 'react'
import Drawer from './Drawer'
import { Meta, Story } from '@storybook/react'
import styled from 'styled-components'

export default {
  title: 'Components/Drawer',
  component: Drawer,
} as Meta<React.ComponentProps<typeof Drawer>>

const Template: Story = (args) => {
  return (
    <StoryWrapper>
      <Drawer
        title={args.title}
        width={args.width}
        color={args.color}
        background={args.background}
      >
        Test
        <DrawerContentMock />
        Test2
        <DrawerContentMock />
      </Drawer>
      Some other content
    </StoryWrapper>
  )
}

export const Default = Template.bind({})
Default.args = {
  title: 'DRAWER',
  width: '400px',
  color: 'white',
  background: '#7BB79B',
}

export const NoParams: Story = () => {
  return (
    <StoryWrapper>
      <Drawer>
        Test
        <DrawerContentMock />
        Test2
        <DrawerContentMock />
      </Drawer>
      Some other content
    </StoryWrapper>
  )
}

export const Larger: Story = () => {
  return (
    <StoryLarge>
      <Drawer
        title="LARGE DRAWER"
        width="600px"
        color="white"
        background="#7BB79B"
      >
        Test
        <DrawerContentMock />
        Test2
        <DrawerContentMock />
      </Drawer>
      Some other content
    </StoryLarge>
  )
}

const StoryWrapper = styled.div`
  font-family: Roboto;
  font-size: 1em;
  max-width: 500px;
`

const StoryLarge = styled.div`
  font-family: Roboto;
  font-size: 2em;
`

const DrawerContentMock = styled.div`
  min-height: 200px;
  background: magenta;
`
