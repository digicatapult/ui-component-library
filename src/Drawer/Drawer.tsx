import React, {
  PropsWithChildren,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react'
import styled from 'styled-components'

import Grid from '../Grid'

export interface DrawerProps {
  defaultIsOpen?: boolean
  title?: string
  width?: string
  color?: string
  background?: string
}

const Drawer: React.FC<PropsWithChildren<DrawerProps>> = (props) => {
  const [isOpen, setIsOpen] = useState(props.defaultIsOpen || false)
  const contentRef: React.RefObject<HTMLDivElement> | null = useRef(null)

  const clickHandler = useCallback(() => {
    setIsOpen(!isOpen)
  }, [isOpen, setIsOpen])

  const title = props.title || ''
  const width = props.width || '100%'
  const color = props.color || 'black'
  const background = props.background || 'transparent'
  const contentHeight = contentRef.current?.clientHeight || 1000

  return (
    <DrawerWrapper width={width}>
      <DrawerHeader
        title={title}
        color={color}
        background={background}
        isOpen={isOpen}
        onClick={clickHandler}
      ></DrawerHeader>
      <DrawerContentOverflow isOpen={isOpen} maxHeight={contentHeight}>
        <DrawContentWrapper ref={contentRef}>
          {props.children}
        </DrawContentWrapper>
      </DrawerContentOverflow>
    </DrawerWrapper>
  )
}

interface DrawerIconProps {
  color: string
}

interface DrawerTitleProps {
  color: string
}

interface DrawIndicatorProps {
  color: string
  isOpen: boolean
}

interface DrawerContainerProps {
  width: string
}

interface DrawerHeaderProps extends React.DOMAttributes<HTMLButtonElement> {
  title: string
  isOpen: boolean
  background: string
  color: string
}

interface DrawerContentProps {
  isOpen: boolean
  maxHeight: number
}

const DrawerHeader: React.FC<DrawerHeaderProps> = (props) => {
  return (
    <DrawHeaderWrapper {...props}>
      <Grid
        areas={[['icon', 'title', 'indicator']]}
        rows={['auto']}
        columns={['auto', '1fr', 'auto']}
        gap="1em"
        alignItems="center"
      >
        <Grid.Panel area="icon">
          <DrawHeaderIcon color={props.color} />
        </Grid.Panel>
        <Grid.Panel area="title" justifySelf={'left'}>
          <DrawerTitle color={props.color}>{props.title}</DrawerTitle>
        </Grid.Panel>
        <Grid.Panel area="indicator">
          <DrawIndicatorIcon color={props.color} isOpen={props.isOpen}>
            â–¼
          </DrawIndicatorIcon>
        </Grid.Panel>
      </Grid>
    </DrawHeaderWrapper>
  )
}

const DrawerWrapper = styled.div<DrawerContainerProps>`
  width: ${({ width }) => width};
`

const DrawHeaderIcon = styled.div<DrawerIconProps>`
  background: ${({ color }) => color};
  height: calc(1em / 16);
  width: 1em;
  margin: 0 calc(1em * 0.35 / 2);

  ::before {
    content: '';
    display: block;
    background: ${({ color }) => color};
    height: calc(1em / 16);
    width: 1em;
    transform: translateY(calc(-4 * 1em / 16)) scaleX(1.35);
  }

  ::after {
    content: '';
    display: block;
    background: ${({ color }) => color};
    height: calc(1em / 16);
    width: 1em;
    transform: translateY(calc(3 * 1em / 16)) scaleX(0.51);
  }
`

const DrawerTitle = styled.div<DrawerTitleProps>`
  color: ${({ color }) => color};
`

const DrawIndicatorIcon = styled('div')<DrawIndicatorProps>`
  margin: 0 0.5em;
  font-size: 0.5em;
  color: ${({ color }) => color};
  transition: transform 0.5s ease;
  transform: scaleX(2) ${({ isOpen }) => (isOpen ? 'scaleY(-1)' : '')};
`

const DrawHeaderWrapper = styled('button')<DrawerHeaderProps>`
  width: 100%;
  font: inherit;
  background: ${({ background }) => background};
  padding: 8px 16px;
  border: 0;
`

// Overflow wrapper to ensure that when we hide the content the translated element
// doesn't overflow and instead hides the content
const DrawerContentOverflow = styled.div<DrawerContentProps>`
  overflow-y: hidden;
  width: 100%;
  transition: max-height 0.5s ease;
  ${({ isOpen, maxHeight }) =>
    isOpen
      ? `
    max-height: ${maxHeight}px;
  `
      : `
    max-height: 0px;
  `}
`

const DrawContentWrapper = styled.div`
  position: relative;
  width: 100%;
`

export default Drawer
