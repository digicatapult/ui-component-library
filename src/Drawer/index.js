import Drawer from './Drawer.tsx'
export default Drawer
