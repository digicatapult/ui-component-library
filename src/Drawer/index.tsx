import React, {
  PropsWithChildren,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react'
import styled from 'styled-components'

import Grid from '../Grid/index.js'

export interface DrawerProps {
  defaultIsOpen?: boolean
  title?: string
  width?: string
  color?: string
  background?: string
}

const Drawer: React.FC<PropsWithChildren<DrawerProps>> = ({
  defaultIsOpen = false,
  title = '',
  width = '100%',
  color = 'black',
  background = 'transparent',
  children,
}) => {
  const [isOpen, setIsOpen] = useState(defaultIsOpen)
  const contentRef: React.RefObject<HTMLDivElement> | null = useRef(null)

  const clickHandler = useCallback(() => {
    setIsOpen(!isOpen)
  }, [isOpen, setIsOpen])

  const contentHeight = contentRef.current?.clientHeight || 1000

  return (
    <DrawerWrapper width={width}>
      <DrawerHeader
        title={title}
        color={color}
        background={background}
        isOpen={isOpen}
        onClick={clickHandler}
      ></DrawerHeader>
      <DrawerContentOverflow isOpen={isOpen} maxHeight={contentHeight}>
        <DrawerContentWrapper ref={contentRef}>{children}</DrawerContentWrapper>
      </DrawerContentOverflow>
    </DrawerWrapper>
  )
}

interface DrawerIconProps {
  color: string
}

interface DrawerTitleProps {
  color: string
}

interface DrawerIndicatorProps {
  color: string
  isOpen: boolean
}

interface DrawerContainerProps {
  width: string
}

interface DrawerHeaderProps extends React.DOMAttributes<HTMLButtonElement> {
  title: string
  isOpen: boolean
  background: string
  color: string
}

interface DrawerContentProps {
  isOpen: boolean
  maxHeight: number
}

const DrawerHeader: React.FC<DrawerHeaderProps> = (props) => {
  return (
    <DrawerHeaderWrapper {...props}>
      <Grid
        areas={[['icon', 'title', 'indicator']]}
        rows={['auto']}
        columns={['auto', '1fr', 'auto']}
        gap="1em"
        alignItems="center"
      >
        <Grid.Panel area="icon">
          <DrawerHeaderIcon color={props.color} />
        </Grid.Panel>
        <Grid.Panel area="title" justifySelf={'left'}>
          <DrawerTitle color={props.color}>{props.title}</DrawerTitle>
        </Grid.Panel>
        <Grid.Panel area="indicator">
          <DrawerIndicatorIcon color={props.color} isOpen={props.isOpen}>
            â–¼
          </DrawerIndicatorIcon>
        </Grid.Panel>
      </Grid>
    </DrawerHeaderWrapper>
  )
}

const DrawerWrapper = styled.div<DrawerContainerProps>`
  width: ${({ width }) => width};
`

const DrawerHeaderIcon = styled.div<DrawerIconProps>`
  background: ${({ color }) => color};
  height: calc(1em / 16);
  width: 1em;
  margin: 0 calc(1em * 0.35 / 2);

  ::before {
    content: '';
    display: block;
    background: ${({ color }) => color};
    height: calc(1em / 16);
    width: 1em;
    transform: translateY(calc(-4 * 1em / 16)) scaleX(1.35);
  }

  ::after {
    content: '';
    display: block;
    background: ${({ color }) => color};
    height: calc(1em / 16);
    width: 1em;
    transform: translateY(calc(3 * 1em / 16)) scaleX(0.51);
  }
`

const DrawerTitle = styled.div<DrawerTitleProps>`
  color: ${({ color }) => color};
`

const DrawerIndicatorIcon = styled('div')<DrawerIndicatorProps>`
  margin: 0 0.5em;
  font-size: 0.5em;
  color: ${({ color }) => color};
  transition: transform 0.5s ease;
  transform: scaleX(2) ${({ isOpen }) => (isOpen ? 'scaleY(-1)' : '')};
`

const DrawerHeaderWrapper = styled('button')<DrawerHeaderProps>`
  width: 100%;
  font: inherit;
  background: ${({ background }) => background};
  padding: 8px 16px;
  border: 0;

  :hover {
    cursor: pointer;
  }
`

// Overflow wrapper to ensure that when we hide the content the translated element
// doesn't overflow and instead hides the content
const DrawerContentOverflow = styled.div<DrawerContentProps>`
  overflow-y: hidden;
  width: 100%;
  transition: max-height 0.5s ease;
  ${({ isOpen, maxHeight }) =>
    isOpen
      ? `
    max-height: ${maxHeight}px;
    animation: 1s fadeIn;
    animation-fill-mode: forwards;
  `
      : `
    max-height: 0px;
  `}

  @keyframes fadeIn {
    99% {
      overflow: hidden;
    }
    100% {
      overflow: visible;
    }
  }
`

const DrawerContentWrapper = styled.div`
  position: relative;
  width: 100%;
`

export default Drawer
