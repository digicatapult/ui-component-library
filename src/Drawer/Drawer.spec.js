import React from 'react'
import renderer, { act } from 'react-test-renderer'
import 'jest-styled-components'

import Drawer from './index.tsx'

import styled from 'styled-components'

describe('Drawer', () => {
  test('simple', () => {
    const tree = renderer
      .create(
        <Drawer>
          <DrawerContentMock />
        </Drawer>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('onClick', async () => {
    const instance = renderer.create(
      <Drawer>
        <DrawerContentMock />
      </Drawer>
    )

    await act(() => {
      instance.root.findByType('button').props.onClick()
    })

    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with overrides', () => {
    const tree = renderer
      .create(
        <Drawer title="FILTER" width="300px" background="magenta">
          <DrawerContentMock />
        </Drawer>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

const DrawerContentMock = styled.div`
  min-height: 500px;
  background: cyan;
`
