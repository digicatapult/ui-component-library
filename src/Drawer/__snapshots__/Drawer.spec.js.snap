// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer onClick 1`] = `
.c3 {
  grid-area: icon;
}

.c5 {
  grid-area: title;
  justify-self: left;
}

.c7 {
  grid-area: indicator;
}

.c2 {
  display: grid;
  grid-template-areas: "icon title indicator";
  grid-template-rows: auto;
  grid-template-columns: auto 1fr auto;
  gap: 1em;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 {
  width: 100%;
}

.c4 {
  background: black;
  height: calc(1em / 16);
  width: 1em;
  margin: 0 calc(1em * 0.35 / 2);
}

.c4::before {
  content: '';
  display: block;
  background: black;
  height: calc(1em / 16);
  width: 1em;
  -webkit-transform: translateY(calc(-4 * 1em / 16)) scaleX(1.35);
  -ms-transform: translateY(calc(-4 * 1em / 16)) scaleX(1.35);
  transform: translateY(calc(-4 * 1em / 16)) scaleX(1.35);
}

.c4::after {
  content: '';
  display: block;
  background: black;
  height: calc(1em / 16);
  width: 1em;
  -webkit-transform: translateY(calc(3 * 1em / 16)) scaleX(0.51);
  -ms-transform: translateY(calc(3 * 1em / 16)) scaleX(0.51);
  transform: translateY(calc(3 * 1em / 16)) scaleX(0.51);
}

.c6 {
  color: black;
}

.c8 {
  margin: 0 0.5em;
  font-size: 0.5em;
  color: black;
  -webkit-transition: -webkit-transform 0.5s ease;
  -webkit-transition: transform 0.5s ease;
  transition: transform 0.5s ease;
  -webkit-transform: scaleX(2) scaleY(-1);
  -ms-transform: scaleX(2) scaleY(-1);
  transform: scaleX(2) scaleY(-1);
}

.c1 {
  width: 100%;
  font: inherit;
  background: transparent;
  padding: 8px 16px;
  border: 0;
}

.c1:hover {
  cursor: pointer;
}

.c9 {
  overflow-y: hidden;
  width: 100%;
  -webkit-transition: max-height 0.5s ease;
  transition: max-height 0.5s ease;
  max-height: 1000px;
}

.c10 {
  position: relative;
  width: 100%;
}

.c11 {
  min-height: 500px;
  background: cyan;
}

<div
  className="c0"
  width="100%"
>
  <button
    className="c1"
    color="black"
    onClick={[Function]}
    title=""
  >
    <div
      className="c2"
      rows={
        [
          "auto",
        ]
      }
    >
      <div
        className="c3"
      >
        <div
          className="c4"
          color="black"
        />
      </div>
      <div
        className="c5"
      >
        <div
          className="c6"
          color="black"
        />
      </div>
      <div
        className="c7"
      >
        <div
          className="c8"
          color="black"
        >
          ▼
        </div>
      </div>
    </div>
  </button>
  <div
    className="c9"
  >
    <div
      className="c10"
    >
      <div
        className="c11"
      />
    </div>
  </div>
</div>
`;

exports[`Drawer simple 1`] = `
.c3 {
  grid-area: icon;
}

.c5 {
  grid-area: title;
  justify-self: left;
}

.c7 {
  grid-area: indicator;
}

.c2 {
  display: grid;
  grid-template-areas: "icon title indicator";
  grid-template-rows: auto;
  grid-template-columns: auto 1fr auto;
  gap: 1em;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 {
  width: 100%;
}

.c4 {
  background: black;
  height: calc(1em / 16);
  width: 1em;
  margin: 0 calc(1em * 0.35 / 2);
}

.c4::before {
  content: '';
  display: block;
  background: black;
  height: calc(1em / 16);
  width: 1em;
  -webkit-transform: translateY(calc(-4 * 1em / 16)) scaleX(1.35);
  -ms-transform: translateY(calc(-4 * 1em / 16)) scaleX(1.35);
  transform: translateY(calc(-4 * 1em / 16)) scaleX(1.35);
}

.c4::after {
  content: '';
  display: block;
  background: black;
  height: calc(1em / 16);
  width: 1em;
  -webkit-transform: translateY(calc(3 * 1em / 16)) scaleX(0.51);
  -ms-transform: translateY(calc(3 * 1em / 16)) scaleX(0.51);
  transform: translateY(calc(3 * 1em / 16)) scaleX(0.51);
}

.c6 {
  color: black;
}

.c8 {
  margin: 0 0.5em;
  font-size: 0.5em;
  color: black;
  -webkit-transition: -webkit-transform 0.5s ease;
  -webkit-transition: transform 0.5s ease;
  transition: transform 0.5s ease;
  -webkit-transform: scaleX(2);
  -ms-transform: scaleX(2);
  transform: scaleX(2);
}

.c1 {
  width: 100%;
  font: inherit;
  background: transparent;
  padding: 8px 16px;
  border: 0;
}

.c1:hover {
  cursor: pointer;
}

.c9 {
  overflow-y: hidden;
  width: 100%;
  -webkit-transition: max-height 0.5s ease;
  transition: max-height 0.5s ease;
  max-height: 0px;
}

.c10 {
  position: relative;
  width: 100%;
}

.c11 {
  min-height: 500px;
  background: cyan;
}

<div
  className="c0"
  width="100%"
>
  <button
    className="c1"
    color="black"
    onClick={[Function]}
    title=""
  >
    <div
      className="c2"
      rows={
        [
          "auto",
        ]
      }
    >
      <div
        className="c3"
      >
        <div
          className="c4"
          color="black"
        />
      </div>
      <div
        className="c5"
      >
        <div
          className="c6"
          color="black"
        />
      </div>
      <div
        className="c7"
      >
        <div
          className="c8"
          color="black"
        >
          ▼
        </div>
      </div>
    </div>
  </button>
  <div
    className="c9"
  >
    <div
      className="c10"
    >
      <div
        className="c11"
      />
    </div>
  </div>
</div>
`;

exports[`Drawer with overrides 1`] = `
.c3 {
  grid-area: icon;
}

.c5 {
  grid-area: title;
  justify-self: left;
}

.c7 {
  grid-area: indicator;
}

.c2 {
  display: grid;
  grid-template-areas: "icon title indicator";
  grid-template-rows: auto;
  grid-template-columns: auto 1fr auto;
  gap: 1em;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 {
  width: 300px;
}

.c4 {
  background: black;
  height: calc(1em / 16);
  width: 1em;
  margin: 0 calc(1em * 0.35 / 2);
}

.c4::before {
  content: '';
  display: block;
  background: black;
  height: calc(1em / 16);
  width: 1em;
  -webkit-transform: translateY(calc(-4 * 1em / 16)) scaleX(1.35);
  -ms-transform: translateY(calc(-4 * 1em / 16)) scaleX(1.35);
  transform: translateY(calc(-4 * 1em / 16)) scaleX(1.35);
}

.c4::after {
  content: '';
  display: block;
  background: black;
  height: calc(1em / 16);
  width: 1em;
  -webkit-transform: translateY(calc(3 * 1em / 16)) scaleX(0.51);
  -ms-transform: translateY(calc(3 * 1em / 16)) scaleX(0.51);
  transform: translateY(calc(3 * 1em / 16)) scaleX(0.51);
}

.c6 {
  color: black;
}

.c8 {
  margin: 0 0.5em;
  font-size: 0.5em;
  color: black;
  -webkit-transition: -webkit-transform 0.5s ease;
  -webkit-transition: transform 0.5s ease;
  transition: transform 0.5s ease;
  -webkit-transform: scaleX(2);
  -ms-transform: scaleX(2);
  transform: scaleX(2);
}

.c1 {
  width: 100%;
  font: inherit;
  background: magenta;
  padding: 8px 16px;
  border: 0;
}

.c1:hover {
  cursor: pointer;
}

.c9 {
  overflow-y: hidden;
  width: 100%;
  -webkit-transition: max-height 0.5s ease;
  transition: max-height 0.5s ease;
  max-height: 0px;
}

.c10 {
  position: relative;
  width: 100%;
}

.c11 {
  min-height: 500px;
  background: cyan;
}

<div
  className="c0"
  width="300px"
>
  <button
    className="c1"
    color="black"
    onClick={[Function]}
    title="FILTER"
  >
    <div
      className="c2"
      rows={
        [
          "auto",
        ]
      }
    >
      <div
        className="c3"
      >
        <div
          className="c4"
          color="black"
        />
      </div>
      <div
        className="c5"
      >
        <div
          className="c6"
          color="black"
        >
          FILTER
        </div>
      </div>
      <div
        className="c7"
      >
        <div
          className="c8"
          color="black"
        >
          ▼
        </div>
      </div>
    </div>
  </button>
  <div
    className="c9"
  >
    <div
      className="c10"
    >
      <div
        className="c11"
      />
    </div>
  </div>
</div>
`;
